feat: enhance admin api filter (#12291)
apache/apisix: docs/en/latest/admin-api.md

}
```

### Support reference filtering query

:::note

> This feature was introduced in APISIX 3.13.0.

Reviewer: Don't think this sentence is necessary? This doc change will go into "next" and only become visible in 3.13.0 version once released.

I think we can get rid of this note and just put the sentence below in a regular paragraph.

> APISIX supports querying routes and stream routes by `service_id` and `upstream_id`. Other resources or fields are not currently supported.

WDYT?

Author: I did encounter this. Users do not take care to manually select the older version of the documentation they are using, and trying a particular new version of a supported feature on an older version directly from the documentation.

This is confusing.

> It now supports querying by `service_id` and `upstream_id` on routes and stream routes. Other resources or other fields are not supported.

Reviewer: ```suggestion
APISIX supports querying routes and stream routes by `service_id` and `upstream_id`. Other resources or fields are not currently supported.
```

others LGTM

:::

When getting a list of resources, it supports a `filter` for filtering resources by filters.

It is encoded in the following manner.

```text

> filter=escape_uri(key1=value2&key2=value2)

Reviewer 2: ```suggestion
filter=escape_uri(key1=value1&key2=value2)
```

Author: Fixed

```

> The following example filters routes using `service_id`. When multiple filters are input at the same time, the result will be the intersection of them.

Reviewer: ```suggestion
The following example filters routes using `service_id`. Applying multiple filters simultaneously will return results that match all filter conditions.
```

```shell

> curl 'http://127.0.0.1:9180/apisix/admin/routes?filter=service_id%3D1' \

Reviewer 3: one question: can we adjust to use the format `a[b]=c&a[d]=e`? It seems more common. https://www.npmjs.com/package/qs

Author: There is no existing API for parsing querystring for such objects in openresty, so you may have to use regular expressions to do so.
I don't want to do that right now. The existing format can be implemented in any language client, it's nothing more than a URI-encoded string.

```lua
ngx.log(ngx.ERR, require("cjson").encode(ngx.decode_args("filter%5Bservice_id%5D=a&filter%5Bservice_id%5D=b&b=c&c=c")))
```

```text
[lua] test.lua:1: {"filter[service_id]":["a","b"],"b":"c","c":"c"}, context: ngx.timer
```

Author: ![image](https://github.com/user-attachments/assets/72f1f525-5693-4a85-b8c7-2c578ef8308e)

It's too complicated to support something like this.
